core/src/main/java/org/apache/iceberg/RollingManifestWriter.java

```diff
@@ -0,0 +1,243 @@
+   * @param addedFile a data file
+   */
+  @Override
+  public void add(T addedFile) {
+    currentWriter.add(addedFile);
+    tryRollingToNewFile();
+  }
+
+  /**
+   * Add an added entry for a file with a specific sequence number.
+   *
+   * <p>The entry's snapshot ID will be this manifest's snapshot ID. The entry's data sequence
+   * number will be the provided data sequence number. The entry's file sequence number will be
+   * assigned at commit.
+   *
+   * @param addedFile a data file
+   * @param dataSequenceNumber a data sequence number for the file
+   */
+  public void add(T addedFile, long dataSequenceNumber) {
+    currentWriter.add(addedFile, dataSequenceNumber);
+    tryRollingToNewFile();
+  }
+
+  /**
+   * Add an existing entry for a file.
+   *
+   * <p>The original data and file sequence numbers, snapshot ID, which were assigned at commit,
+   * must be preserved when adding an existing entry.
+   *
+   * @param existingFile a file
+   * @param fileSnapshotId snapshot ID when the data file was added to the table
+   * @param dataSequenceNumber a data sequence number of the file (assigned when the file was added)
+   * @param fileSequenceNumber a file sequence number (assigned when the file was added)
+   */
+  public void existing(
+      T existingFile, long fileSnapshotId, long dataSequenceNumber, Long fileSequenceNumber) {
+    currentWriter.existing(existingFile, fileSnapshotId, dataSequenceNumber, fileSequenceNumber);
+    tryRollingToNewFile();
+  }
+
+  /**
+   * Add a delete entry for a file.
+   *
+   * <p>The entry's snapshot ID will be this manifest's snapshot ID. However, the original data and
+   * file sequence numbers of the file must be preserved when the file is marked as deleted.
+   *
+   * @param deletedFile a file
+   * @param dataSequenceNumber a data sequence number of the file (assigned when the file was added)
+   * @param fileSequenceNumber a file sequence number (assigned when the file was added)
+   */
+  public void delete(T deletedFile, long dataSequenceNumber, Long fileSequenceNumber) {
+    currentWriter.delete(deletedFile, dataSequenceNumber, fileSequenceNumber);
+    tryRollingToNewFile();
+  }
+
+  private void tryRollingToNewFile() {
+    currentFileRows++;
+
+    if (currentWriter.length() > targetFileSizeInBytes) {
+      closeCurrentWriter();
+      openCurrentWriter();
+    }
+  }
+
+  private void openCurrentWriter() {
+    Preconditions.checkState(currentWriter == null, "Current writer has been already initialized");
+
+    this.currentFile = outputFileFactory.newManifestOutput();
+    this.currentFileRows = 0;
+    this.currentWriter = newManifestWriter(formatVersion, snapshotId, spec, currentFile);
+  }
+
+  private void closeCurrentWriter() {
+    if (currentWriter != null) {
+      try {
+        currentWriter.close();
+      } catch (IOException e) {
+        throw new UncheckedIOException("Failed to close current writer", e);
+      }
+
+      if (currentFileRows == 0L) {
```

Comment:
I feel like this class can be simplified by not dealing with `FileIO` and manifest creation. Also, I am not sure there is a lot of value in implementing `FileAppender` given that we don't implement length and metrics (I don't think we should). What about something like this?

```
public class RollingManifestWriter<F extends ContentFile<F>> implements Closeable {

  private static final int ROWS_DIVISOR = SOME_VALUE;

  private final Supplier<ManifestWriter<F>> manifestWriterSupplier;
  private final long targetFileSizeInBytes;
  private final List<ManifestFile> manifestFiles;

  private long currentFileRows = 0;
  private ManifestWriter<F> currentWriter = null;

  private boolean closed = false;

  public RollingManifestWriter(
      Supplier<ManifestWriter<F>> manifestWriterSupplier,
      long targetFileSizeInBytes) {
    this.manifestWriterSupplier = manifestWriterSupplier;
    this.targetFileSizeInBytes = targetFileSizeInBytes;
    this.manifestFiles = Lists.newArrayList();
  }

  public void add(F addedFile) {
    currentWriter().add(addedFile);
    currentFileRows++;
  }

  ...

  private ManifestWriter<F> currentWriter() {
    if (currentWriter == null) {
      this.currentWriter = manifestWriterSupplier.get();
    } else if (shouldRollToNewFile()) {
      closeCurrentWriter();
      this.currentWriter = manifestWriterSupplier.get();
    }

    return currentWriter;
  }

  private boolean shouldRollToNewFile() {
    return currentFileRows % ROWS_DIVISOR == 0 && currentWriter.length() >= targetFileSizeInBytes;
  }
}
```

By migrating to a lazy writer, we don't have to check if the current file is empty and delete it on close. It should be enough to have a manifest writer supplier and the target manifest size. That way, this class won't need to change if we change the signature of how the underlying manifest writers are created. 

Calling this in `SnapshotProducer` would be trivial.

```
protected RollingManifestWriter<DataFile> newRollingManifestWriter(PartitionSpec spec) {
  return new RollingManifestWriter<>(() -> newManifestWriter(spec), targetManifestSize);
}

protected RollingManifestWriter<DeleteFile> newRollingDeleteManifestWriter(PartitionSpec spec) {
  return new RollingManifestWriter<>(() -> newDeleteManifestWriter(spec), targetManifestSize);
}
```

We may skip changes in `ManifestFiles` and drop `ManifestOutputFileFactory` for now.

Reply:
Much cleaner implementation. Updated.